### PR TITLE
Remove failure if version not present in the reported properties response

### PR DESCRIPTION
**Pull request by momuno**
Remove the failure if version is not present in the reported properties response. 
It will be present from IoT Hub, but absent from IoT Edge.

### CODE REVIEW

File: sdk/src/azure/iot/az_iot_hub_client_twin.c [hunk 1]
```diff
@@ -173,8 +173,9 @@ AZ_NODISCARD az_result az_iot_hub_client_twin_parse_received_topic(
       {
         // Is a reported prop response
         out_response->response_type = AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_REPORTED_PROPERTIES;
-        _az_RETURN_IF_FAILED(az_iot_message_properties_find(
-            &props, az_iot_hub_twin_version_prop, &out_response->version));
+        out_response->version = AZ_SPAN_EMPTY;
+        az_iot_message_properties_find(
+            &props, az_iot_hub_twin_version_prop, &out_response->version);
```
Comment (danewalton) [inline sdk/src/azure/iot/az_iot_hub_client_twin.c:178]: We should still check if this fails, and if it does with a `AZ_ERROR_ITEM_NOT_FOUND`, explicitly set `&out_response->version = AZ_SPAN_EMPTY`. Wouldn't want garbage going in there by accident.

Comment (danewalton) [reply sdk/src/azure/iot/az_iot_hub_client_twin.c:178]: Also...you're going to hate me.....a test with this scenario :)

Comment (momuno) [reply sdk/src/azure/iot/az_iot_hub_client_twin.c:178]: Thanks for commenting!  I was curious to know what others were thinking on this issue.  We will need to do a test with IoT Edge, not IoT Hub.  Do you know how we go about creating devices that talk to IoT Edge and not Hub?

Comment (momuno) [reply sdk/src/azure/iot/az_iot_hub_client_twin.c:178]: That makes sense to switch the order to look only for the specific AZ failure of version being absent.

Comment (danewalton) [reply sdk/src/azure/iot/az_iot_hub_client_twin.c:178]: > Thanks for commenting! I was curious to know what others were thinking on this issue. We will need to do a test with IoT Edge, not IoT Hub. Do you know how we go about creating devices that talk to IoT Edge and not Hub?

A unit test should be enough for it.

Comment (momuno) [reply sdk/src/azure/iot/az_iot_hub_client_twin.c:178]: Side note: The underlying az_iot_message_properties_find only returns AZ_OK or AZ_ERROR_ITEM_NOT_FOUND, so for now there isn't another failure option.  However, if the logic ever changed in the future to include a different possible error return, we would be missing that scenario if we don't add the check in at this level.

**Review by CIPop — CHANGES_REQUESTED** [review]
LGTM. Please add a UT to pin down this new behavior.

**Comment by CIPop** [comment]
While not an API breaking change, this is a behavior breaking change: apps relying on finding a version if the parser passed will now find an AZ_SPAN_EMPTY instead. We already know that devices connecting to Edge will fail so I believe the risk is very small. @ahsonkhan @RickWinter FYI.

**Comment by momuno** [comment]
> While not an API breaking change, this is a behavior breaking change: apps relying on finding a version if the parser passed will now find an AZ_SPAN_EMPTY instead. We already know that devices connecting to Edge will fail so I believe the risk is very small. @ahsonkhan @RickWinter FYI.

I agree Cris and was thinking about this as well.  This all works if the Hub does not stop sending version in the topic.  But if they stop, it could break apps that are expecting version and are only checking for AZ_OK.

**Comment by momuno** [comment]
@CIPop @danewalton UT added.

File: sdk/tests/iot/hub/test_az_iot_hub_client_twin.c [hunk 2]
```diff
@@ -26,16 +26,18 @@ static const az_span test_device_request_id = AZ_SPAN_LITERAL_FROM_STR("id_one")
 static const az_span test_twin_received_topic_desired_success
     = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/PATCH/properties/desired/?$version=id_one");
 static const az_span test_twin_received_topic_fail
-    = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/rez/200");
+    = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/rez/200"); // rez DNE
 static const az_span test_twin_received_topic_incomplete_fail
-    = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/res/200");
+    = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/res/200"); // Missing rid
```
Comment (danewalton) [inline sdk/tests/iot/hub/test_az_iot_hub_client_twin.c:31]: great comments thanks for these!

**Review by danewalton — APPROVED** [review]

File: sdk/tests/iot/hub/test_az_iot_hub_client_twin.c (hunk 2 above)
Comment (momuno) [reply sdk/tests/iot/hub/test_az_iot_hub_client_twin.c:31]: aw thanks dane. Glad it helps.

File: sdk/tests/iot/hub/test_az_iot_hub_client_twin.c [hunk 3]
```diff
@@ -26,16 +26,18 @@ static const az_span test_device_request_id = AZ_SPAN_LITERAL_FROM_STR("id_one")
 static const az_span test_twin_received_topic_desired_success
     = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/PATCH/properties/desired/?$version=id_one");
 static const az_span test_twin_received_topic_fail
-    = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/rez/200");
+    = AZ_SPAN_LITERAL_FROM_STR("$iothub/twin/rez/200"); // rez DNE
```
Comment (CIPop) [inline sdk/tests/iot/hub/test_az_iot_hub_client_twin.c:29]: What does DNE mean?

**Review by CIPop — APPROVED** [review]

File: sdk/tests/iot/hub/test_az_iot_hub_client_twin.c (hunk 3 above)
Comment (momuno) [reply sdk/tests/iot/hub/test_az_iot_hub_client_twin.c:29]: Does Not Exist.

Comment (momuno) [reply sdk/tests/iot/hub/test_az_iot_hub_client_twin.c]: Expanded :)

File: sdk/src/azure/iot/az_iot_hub_client_twin.c [hunk 4]
```diff
@@ -173,8 +173,17 @@ AZ_NODISCARD az_result az_iot_hub_client_twin_parse_received_topic(
...
+        result = az_iot_message_properties_find(
+            &props, az_iot_hub_twin_version_prop, &out_response->version);
+        if (result == AZ_ERROR_ITEM_NOT_FOUND)
+        {
+          out_response->version = AZ_SPAN_EMPTY;
+        }
+        else
+        {
+          _az_RETURN_IF_FAILED(result);
+        }
```
Comment (ahsonkhan) [inline sdk/src/azure/iot/az_iot_hub_client_twin.c:186]: Are we sure this is what we want? We are changing `AZ_ERROR_ITEM_NOT_FOUND` return value to `AZ_OK` on line 195 now.

Do we need to update the docs here:
https://github.com/Azure/azure-sdk-for-c/blob/cc7beafdb56f2f495d488451ca2f28f855070a17/sdk/inc/azure/iot/az_iot_hub_client.h#L625-L632

Comment (momuno) [reply sdk/src/azure/iot/az_iot_hub_client_twin.c:186]: We were notified via a colleague on Azure RTOS who had a customer issue.  Through debugging, it was found that IoT Edge did not supply the $version in the reported properties response while IoT Hub did.  This was confirmed by IoT Edge.

Can you please remind me where I create the changelog entry? Thanks!

Comment (ahsonkhan) [reply sdk/src/azure/iot/az_iot_hub_client_twin.c:186]: > Can you please remind me where I create the changelog entry? Thanks!

Here:
https://github.com/Azure/azure-sdk-for-c/blob/main/CHANGELOG.md#bugs-fixed
